x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/BlobSpeedTestAction.java

```diff
@@ -0,0 +1,784 @@
+                    );
+                } else {
+                    // no need for extra synchronization after checking if we were cancelled a couple of lines ago -- we haven't notified
+                    // the outer listener yet so any bans on the children are still in place
+                    transportService.sendChildRequest(
+                        node,
+                        GetBlobChecksumAction.NAME,
+                        getBlobChecksumRequest(),
+                        task,
+                        TransportRequestOptions.EMPTY,
+                        new ActionListenerResponseHandler<>(
+                            readNodesListener.map(r -> makeNodeResponse(node, beforeWriteComplete, r)),
+                            GetBlobChecksumAction.Response::new
+                        )
+                    );
+                }
+            }
+        }
+
+        private GetBlobChecksumAction.Request getBlobChecksumRequest() {
+            return new GetBlobChecksumAction.Request(
+                request.getRepositoryName(),
+                request.getBlobPath(),
+                request.getBlobName(),
+                checksumStart,
+                checksumWholeBlob ? 0L : checksumEnd
+            );
+        }
+
+        private NodeResponse makeNodeResponse(DiscoveryNode node, boolean beforeWriteComplete, GetBlobChecksumAction.Response response) {
+            logger.trace(
+                "received read response [{}] from [{}] for [{}] [beforeWriteComplete={}]",
+                response,
+                node,
+                request.getDescription(),
+                beforeWriteComplete
+            );
+            return new NodeResponse(node, beforeWriteComplete, response);
+        }
+
+        private void cancelReadsCleanUpAndReturnFailure(Exception exception) {
+            transportService.getTaskManager().cancelTaskAndDescendants(task, "task failed", false, ActionListener.wrap(() -> {}));
+            cleanUpAndReturnFailure(exception);
+        }
+
+        private void cleanUpAndReturnFailure(Exception exception) {
+            logger.trace(new ParameterizedMessage("speed test failed [{}] cleaning up", request.getDescription()), exception);
+            try {
+                blobContainer.deleteBlobsIgnoringIfNotExists(List.of(request.blobName));
+            } catch (IOException ioException) {
+                exception.addSuppressed(ioException);
+                logger.warn(
+                    new ParameterizedMessage(
+                        "failure during post-failure cleanup while speed-testing repository [{}], you may need to manually remove [{}/{}]",
+                        request.getRepositoryName(),
+                        request.getBlobPath(),
+                        request.getBlobName()
+                    ),
+                    exception
+                );
+            }
+            listener.onFailure(exception);
```

Comment:
TODO: wrap this exception in another `RepositoryVerificationException` to report the details of the write request too.

Reply:
Done in https://github.com/elastic/elasticsearch/pull/67247/commits/64c220d42344fa68a74afb05261a2efcfe82f4e7.